machine/models/baseModel.py

```diff
@@ -0,0 +1,49 @@
+import torch.nn as nn
+import torch.nn.functional as F
+
+class Model(nn.Module):
+    """ 
+    Abstract base class for models.
+
+    Args:
+        encoder_module (baseRNN) :  module that encodes inputs
+        decoder_module (baseRNN, optional):   module to decode encoded inputs
+        decode_function (func, optional): function to generate symbols from output hidden states (default: F.log_softmax)
+
+    """
+
+    def __init__(self, encoder_module, decoder_module=None, decode_function=F.log_softmax):
+        super(Model, self).__init__()
+        self.encoder_module = encoder_module
+        self.decoder_module = decoder_module
+        self.decode_function = decode_function
+
+    def flatten_parameters(self):
+        self.encoder_module.rnn.flatten_parameters()
+        self.decoder_module.rnn.flatten_parameters()
```

Comment:
Of course a child class can override this method, but it again assumes the encoder and decoder to be (single) RNN's.

Reply:
Good point, I suppose this should be an abstract method also.

Reply:
In my base model I defined it like that :
```
for child in self.children():
            try:
                child.flatten_parameters()
            except (AttributeError, NotImplementedError):
                pass
```

It might not be the cleanest method but it's defintely the quickest and least error prone (basically it will flatten parameters of all rnn modules of children classes). As we are focusing on research I think the more we can have the base library do stuff for us the least we have to think in each projects => the best. 

Note that if a child has an object O which doesn't inherit from the base class then it would have to manually flattent these parameters (or just define `flatten_parameters` in O)

Thoughts ?

Reply:
I think this would be a good solution! 
I don't understand your last point though.

And wouldn't `.modules()` be better to iterate over all modules? (https://discuss.pytorch.org/t/module-children-vs-module-modules/4551/4)

Reply:
The last point is not too important I'm just saying that this only works if the RNN is a child of a module class (hard to explain like that but will most often be the case). 

COncerning `.module()` or `.child()` it depends how you do it. In this case (i.e if you define the code in the BaseModel) then it c^should be`.module()`. In my code (and I thik it would be nice to do the same here) I defined a BaseModule (that enherits from `nn.Module`) from which every module in machine would enherit (i.ewrapper around `nn.Module` that adds our helper methods). In the latter case it should be `.child()` as each `BaseModule` object would call `flatten_parameters` thus doing the recusrion for you (i.e with `.child()` you would loop an exponential amount of time over the leaf nodes).

ALl my comments are actually more for a `BaseModule`

Reply:
This should indeed be implemented, but it can also be done later, for now I'll just assume it will be implemented in the subclass and make it an abstract method. There is a "notImplementedError" that says that a generic version should be implemented.